Protobuf library does not compile under gcc with -Wall -Werror

**What version of protobuf and what language are you using?**
Version: 'bc799d78f81115940eec953e2937245c70e3e6e4', # version 3.20.0
Language: C++

**What operating system (Linux, Windows, ...) and version?**

Linux Debian 5.15.15-1rodete2 (2022-02-23)

**What runtime / compiler are you using (e.g., python version or gcc version)**
gcc (-v output below)
```
Using built-in specs.
COLLECT_GCC=gcc
COLLECT_LTO_WRAPPER=/usr/lib/gcc/x86_64-linux-gnu/11/lto-wrapper
OFFLOAD_TARGET_NAMES=nvptx-none:amdgcn-amdhsa
OFFLOAD_TARGET_DEFAULT=1
Target: x86_64-linux-gnu
Configured with: ../src/configure -v --with-pkgversion='Debian 11.2.0-16' --with-bugurl=file:///usr/share/doc/gcc-11/README.Bugs --enable-languages=c,ada,c++,go,brig,d,fortran,objc,obj-c++,m2 --prefix=/usr --with-gcc-major-version-only --program-suffix=-11 --program-prefix=x86_64-linux-gnu- --enable-shared --enable-linker-build-id --libexecdir=/usr/lib --without-included-gettext --enable-threads=posix --libdir=/usr/lib --enable-nls --enable-bootstrap --enable-clocale=gnu --enable-libstdcxx-debug --enable-libstdcxx-time=yes --with-default-libstdcxx-abi=new --enable-gnu-unique-object --disable-vtable-verify --enable-plugin --enable-default-pie --with-system-zlib --enable-libphobos-checking=release --with-target-system-zlib=auto --enable-objc-gc=auto --enable-multiarch --disable-werror --enable-cet --with-arch-32=i686 --with-abi=m64 --with-multilib-list=m32,m64,mx32 --enable-multilib --with-tune=generic --enable-offload-targets=nvptx-none=/build/gcc-11-q5nAlV/gcc-11-11.2.0/debian/tmp-nvptx/usr,amdgcn-amdhsa=/build/gcc-11-q5nAlV/gcc-11-11.2.0/debian/tmp-gcn/usr --without-cuda-driver --enable-checking=release --build=x86_64-linux-gnu --host=x86_64-linux-gnu --target=x86_64-linux-gnu --with-build-config=bootstrap-lto-lean --enable-link-serialization=8
Thread model: posix
Supported LTO compression algorithms: zlib zstd
gcc version 11.2.0 (Debian 11.2.0-16) 
```

**What did you do?**
Attempt to compile library

**What did you expect to see**
Compilation is successful

**What did you see instead?**
```
[564/1020] CXX obj/third_party/protobuf/src/src/google/protobuf/protobuf_full.generated_message_tctable_lite.o
FAILED: obj/third_party/protobuf/src/src/google/protobuf/protobuf_full.generated_message_tctable_lite.o 
g++ -MMD -MF obj/third_party/protobuf/src/src/google/protobuf/protobuf_full.generated_message_tctable_lite.o.d -DHAVE_ZLIB -D_DEBUG -DOS_LINUX -DGOOGLE_PROTOBUF_NO_RTTI -DGOOGLE_PROTOBUF_NO_STATIC_INITIALIZER -DHAVE_PTHREAD -I../third_party/zlib/src -I../third_party/protobuf/src/src -Wall -Werror -fno-strict-aliasing -Wno-maybe-uninitialized -fvisibility=hidden -fPIC -m64 -O0 -g3 -fstack-protector-strong -std=c++14 -fno-exceptions -fno-unwind-tables -fno-asynchronous-unwind-tables -fno-rtti -Wno-sign-compare -Wno-deprecated-declarations -Wno-unused-function -Wno-stringop-overflow -Wno-array-bounds -c ../third_party/protobuf/src/src/google/protobuf/generated_message_tctable_lite.cc -o obj/third_party/protobuf/src/src/google/protobuf/protobuf_full.generated_message_tctable_lite.o
In file included from ../third_party/protobuf/src/src/google/protobuf/generated_message_tctable_lite.cc:36:
../third_party/protobuf/src/src/google/protobuf/generated_message_tctable_impl.h: In function ‘void google::protobuf::internal::AlignFail(uintptr_t) [with long unsigned int align = 4]’:
../third_party/protobuf/src/src/google/protobuf/generated_message_tctable_impl.h:256:1: error: ‘noreturn’ function does return [-Werror]
  256 | }
      | ^
../third_party/protobuf/src/src/google/protobuf/generated_message_tctable_impl.h: In function ‘void google::protobuf::internal::AlignFail(uintptr_t) [with long unsigned int align = 8]’:
../third_party/protobuf/src/src/google/protobuf/generated_message_tctable_impl.h:256:1: error: ‘noreturn’ function does return [-Werror]
cc1plus: all warnings being treated as errors
[607/1020] CXX obj/third_party/protobuf/src/src/google/protobuf/protobuf_lite.generated_message_tctable_lite.o
FAILED: obj/third_party/protobuf/src/src/google/protobuf/protobuf_lite.generated_message_tctable_lite.o 
g++ -MMD -MF obj/third_party/protobuf/src/src/google/protobuf/protobuf_lite.generated_message_tctable_lite.o.d -D_DEBUG -DOS_LINUX -DGOOGLE_PROTOBUF_NO_RTTI -DGOOGLE_PROTOBUF_NO_STATIC_INITIALIZER -DHAVE_PTHREAD -I../third_party/protobuf/src/src -Wall -Werror -fno-strict-aliasing -Wno-maybe-uninitialized -fvisibility=hidden -fPIC -m64 -O0 -g3 -fstack-protector-strong -std=c++14 -fno-exceptions -fno-unwind-tables -fno-asynchronous-unwind-tables -fno-rtti -Wno-sign-compare -Wno-deprecated-declarations -Wno-unused-function -Wno-stringop-overflow -Wno-array-bounds -c ../third_party/protobuf/src/src/google/protobuf/generated_message_tctable_lite.cc -o obj/third_party/protobuf/src/src/google/protobuf/protobuf_lite.generated_message_tctable_lite.o
In file included from ../third_party/protobuf/src/src/google/protobuf/generated_message_tctable_lite.cc:36:
../third_party/protobuf/src/src/google/protobuf/generated_message_tctable_impl.h: In function ‘void google::protobuf::internal::AlignFail(uintptr_t) [with long unsigned int align = 4]’:
../third_party/protobuf/src/src/google/protobuf/generated_message_tctable_impl.h:256:1: error: ‘noreturn’ function does return [-Werror]
  256 | }
      | ^
../third_party/protobuf/src/src/google/protobuf/generated_message_tctable_impl.h: In function ‘void google::protobuf::internal::AlignFail(uintptr_t) [with long unsigned int align = 8]’:
../third_party/protobuf/src/src/google/protobuf/generated_message_tctable_impl.h:256:1: error: ‘noreturn’ function does return [-Werror]
cc1plus: all warnings being treated as errors
[637/1020] CXX obj/third_party/protobuf/src/src/google/protobuf/protobuf_full.descriptor.o
ninja: build stopped: subcommand failed.

```

Make sure you include information that can help us debug (full error message, exception listing, stack trace, logs).

**Anything else we should know about your project / environment**

There doesn't appear to be a -Wno option to override the -Werror behavior for this. This is a hard error and a blocker for compilation.  Is template calling into something that is actually not `[[noreturn]]` on debug builds?

--- a/src/google/protobuf/generated_message_tctable_impl.h
+++ b/src/google/protobuf/generated_message_tctable_impl.h
@@ -32,6 +32,7 @@
 #define GOOGLE_PROTOBUF_GENERATED_MESSAGE_TCTABLE_IMPL_H__
 
 #include <cstdint>
+#include <cstdlib>
 #include <type_traits>
 
 #include <google/protobuf/port.h>
@@ -262,11 +263,11 @@ enum FieldType : uint16_t {
 
 #ifndef NDEBUG
 template <size_t align>
-#ifndef _MSC_VER
-[[noreturn]]
-#endif
 void AlignFail(uintptr_t address) {
   GOOGLE_LOG(FATAL) << "Unaligned (" << align << ") access at " << address;
+
+  // Explicit abort to let compilers know this function does not return
+  abort();
 }
 
 extern template void AlignFail<4>(uintptr_t);

